graylog2-server/src/main/java/org/graylog2/indexer/messages/Messages.java

```diff
@@ -67,48 +81,59 @@ public Messages(Client client, ElasticsearchConfiguration configuration) {
-			throw new DocumentNotFoundException();
-		}
-		
-		return ResultMessage.parseFromSource(r);
-	}
-	
-	public List<String> analyze(String string, String index) throws IndexMissingException {
-		List<String> tokens = Lists.newArrayList();
-		AnalyzeRequestBuilder arb = new AnalyzeRequestBuilder(c.admin().indices(), index, string);
-		AnalyzeResponse r = c.admin().indices().analyze(arb.request()).actionGet();
-		
-		for (AnalyzeToken token : r.getTokens()) {
-			tokens.add(token.getTerm());
-		}
-		
-		return tokens;
-	}
+        GetRequestBuilder grb = new GetRequestBuilder(c, index);
+        grb.setId(messageId);
+
+        GetResponse r = c.get(grb.request()).actionGet();
+
+        if (!r.isExists()) {
+            throw new DocumentNotFoundException();
+        }
+
+        return ResultMessage.parseFromSource(r);
+    }
+
+    public List<String> analyze(String string, String index) throws IndexMissingException {
+        List<String> tokens = Lists.newArrayList();
+        AnalyzeRequestBuilder arb = new AnalyzeRequestBuilder(c.admin().indices(), index, string);
+        AnalyzeResponse r = c.admin().indices().analyze(arb.request()).actionGet();
+
+        for (AnalyzeToken token : r.getTokens()) {
+            tokens.add(token.getTerm());
+        }
+
+        return tokens;
+    }
 
     public boolean bulkIndex(final List<Message> messages) {
         if (messages.isEmpty()) {
             return true;
         }
 
-        final BulkRequestBuilder request = c.prepareBulk();
+        final BulkRequestBuilder requestBuilder = c.prepareBulk().setConsistencyLevel(WriteConsistencyLevel.ONE);
         for (Message msg : messages) {
-            request.add(buildIndexRequest(Deflector.buildName(configuration.getIndexPrefix()),
-                                          msg.toElasticSearchObject(),
-                                          msg.getId())); // Main index.
+            requestBuilder.add(buildIndexRequest(deflectorName, msg.toElasticSearchObject(), msg.getId()));
         }
 
-        request.setConsistencyLevel(WriteConsistencyLevel.ONE);
-
-        final BulkResponse response = c.bulk(request.request()).actionGet();
+        final BulkResponse response;
+        try {
+            response = RetryerBuilder.<BulkResponse>newBuilder()
+                    .retryIfException(ES_TIMEOUT_EXCEPTION_PREDICATE)
+                    .withWaitStrategy(WaitStrategies.exponentialWait(MAX_WAIT_TIME.getQuantity(), MAX_WAIT_TIME.getUnit()))
+                    .build()
+                    .call(new Callable<BulkResponse>() {
+                        @Override
+                        public BulkResponse call() throws Exception {
+                            return c.bulk(requestBuilder.request()).actionGet();
+                        }
+                    });
+        } catch (ExecutionException | RetryException e) {
+            LOG.error("Couldn't bulk index " + messages.size() + " messages.", e);
+            return false;
+        }
```

Comment:
I think it would be good to use the retryer only in the exceptional case like I did in 6623e80. As it is now, we create lots of garbage objects for every bulk request.